File: src/main/java/org/polyfrost/vanillahud/utils/TabListManager.java
package org.polyfrost.vanillahud.utils;

import org.polyfrost.oneconfig.utils.v1.Multithreading;
import org.polyfrost.oneconfig.utils.v1.NetworkUtils;
import com.google.common.collect.ImmutableList;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.mojang.authlib.GameProfile;
import net.minecraft.client.Minecraft;
import net.minecraft.client.network.NetworkPlayerInfo;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

public class TabListManager {
    private static final Gson gson = new GsonBuilder().create();
    private static final Minecraft mc = Minecraft.getMinecraft();
    private static final String devListURL = "https://raw.githubusercontent.com/Polyfrost/VanillaHUD/main/tablist_uuids.json";

    private static final List<String> fallbackDevUUIDs = Arrays.asList(
            "0b4d470f-f2fb-4874-9334-1eaef8ba4804", // DeDiamondPro
            "c8bf4768-af44-48cb-a259-01e42fb7bc79", // ImToggle
            "0e3ee1e0-f4d2-4550-8fe9-4f7a0d2cd08a", // Mixces
            "0d68ec06-ec8f-4558-959f-7a6d7efd7fa5", // Redth
            "a5331404-0e77-440e-8bef-24c071dac1ae" // Wyvest
    );

    public static List<NetworkPlayerInfo> devInfo = ImmutableList.of();

    public static void asyncFetchFallbackList() {
        Multithreading.submit(() -> devInfo = fallbackDevUUIDs.stream()
                .map(UUID::fromString)
                .map(TabListManager::getProfile)
                .map(NetworkPlayerInfo::new)
                .collect(Collectors.toList()));
    }
    public static GameProfile getProfile(UUID uuid) {
        return mc.getSessionService().fillProfileProperties(new GameProfile(uuid, null), true);
    }

    public static void asyncUpdateList() {
        Multithreading.submit(TabListManager::updateList);
    }
    private static void updateList() {
        try {
            String input = NetworkUtils.getString(devListURL);
            if (input == null) return;
            DevList list = gson.fromJson(input, DevList.class);
            devInfo = list.toNetworkInfoList();
        } catch (Exception exception) {
            new RuntimeException("Failed to load list", exception).printStackTrace();
        }
    }

    private static class DevList extends ArrayList<DevList.DevEntry> {
        private static class DevEntry {
            public String name;
            public String id;
            public UUID getUUID() {
                return UUID.fromString(id);
            }
        }

        public List<NetworkPlayerInfo> toNetworkInfoList() {
            return stream()
                .map(DevList.DevEntry::getUUID)
                .map(TabListManager::getProfile)
                .map(NetworkPlayerInfo::new)
                .collect(Collectors.toList());
        }
    }
}

File: src/main/kotlin/org/polyfrost/vanillahud/util/TabListManager.kt
package org.polyfrost.vanillahud.util

import com.mojang.authlib.GameProfile
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import net.minecraft.client.Minecraft
import net.minecraft.client.multiplayer.PlayerInfo
import org.polyfrost.oneconfig.utils.v1.Multithreading
import org.polyfrost.oneconfig.utils.v1.NetworkUtils
import java.util.UUID

object TabListManager {
    private const val DEV_LIST_URL =
        "https://raw.githubusercontent.com/Polyfrost/VanillaHUD/main/tablist_uuids.json"

    private val mc: Minecraft = Minecraft.getInstance()

    private val fallbackUuids = listOf(
        "0b4d470f-f2fb-4874-9334-1eaef8ba4804",
        "c8bf4768-af44-48cb-a259-01e42fb7bc79",
        "0e3ee1e0-f4d2-4550-8fe9-4f7a0d2cd08a",
        "0d68ec06-ec8f-4558-959f-7a6d7efd7fa5",
        "a5331404-0e77-440e-8bef-24c071dac1ae"
    ).map(UUID::fromString)

    @Volatile
    var devInfo: List<PlayerInfo> = emptyList()
        private set

    @Volatile
    private var requested = false

    fun ensureLoaded() {
        if (requested) return
        requested = true
        Multithreading.submit {
            devInfo = fallbackUuids.toPlayerInfoList()
            updateFromNetwork()
        }
    }

    private fun updateFromNetwork() {
        try {
            val input = NetworkUtils.getString(DEV_LIST_URL) ?: return
            val uuids = Json.parseToJsonElement(input).jsonArray
                .map { UUID.fromString(it.jsonObject["id"]!!.jsonPrimitive.content) }
            devInfo = uuids.toPlayerInfoList()
        } catch (e: Exception) {
            RuntimeException("Failed to load VanillaHUD dev list", e).printStackTrace()
        }
    }

    private fun List<UUID>.toPlayerInfoList(): List<PlayerInfo> =
        map { PlayerInfo(getProfile(it), false) }

    private fun getProfile(uuid: UUID): GameProfile =
        try {
            //? if >=1.21.9 {
            mc.services().sessionService.fetchProfile(uuid, true)?.profile() ?: GameProfile(uuid, null)
            //?} else {
            /*mc.minecraftSessionService.fetchProfile(uuid, true)?.profile() ?: GameProfile(uuid, null)
            *///?}
        } catch (_: Exception) {
            GameProfile(uuid, null)
        }
}
